diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -1186,8 +1186,24 @@ int DifferentialFunctionComparator::cmpOperationsWithOperands(
     // convenient to have the code in a separate function.
 
     bool needToCmpOperands = true;
-    if (int Res = cmpOperations(L, R, needToCmpOperands))
+    if (int Res = cmpOperations(L, R, needToCmpOperands)) {
+        auto *CallL = dyn_cast<CallInst>(L);
+        auto *CallR = dyn_cast<CallInst>(R);
+        if (CallL != CallR) {
```

Comment:
While `!=` is logical non-equivalence (XOR) on bool operands, it behaves differently on non-bool operands in case they are two different non-zero objects (logical non-equivalence would return false, `!=` returns true).

Because of this, the code inside the if condition will execute even if both `L` and `R` are call instructions, since the pointers will always be different, which I suppose isn't the desired behavior.

`!!CallL != !!CallR`, `bool(CallL) != bool(CallR)` or simply `(CallL || CallR) && !(CallL && CallR)` should work.

Reply:
Correct, thanks for noting this.

Reply:
Fixed.